wallet/src/actors/storage/mod.rs

```diff
@@ -2,64 +2,97 @@
 //!
 //! It is charge of managing the connection to the key-value database. This actor is blocking so it
 //! must be used with a `SyncArbiter`.
-
+use bincode::deserialize;
+use failure::Error;
+use std::env;
 use std::path::PathBuf;
+use std::sync::Arc;
 
 use actix::prelude::*;
 
-mod error;
+use crate::wallet;
+
+pub mod error;
 mod handlers;
 
-pub use error::Error;
 pub use handlers::*;
 
+/// Expose options for tunning the database.
+pub type Options = rocksdb::Options;
+
 /// Storage actor.
 pub struct Storage {
-    /// Path where to store the wallet database.
-    /// This is used during the wallet generation step.
-    // TODO: remove this allow attribute when the wallet-database is implemented.
-    #[allow(dead_code)]
-    db_path: PathBuf,
-    wallets: Result<rocksdb::DB, rocksdb::Error>,
-    /// It holds the instance of the currently open wallet database.
-    // TODO: remove this allow attribute when the wallet-database is implemented and the client is
-    // able to "open" a wallet.
-    #[allow(dead_code)]
-    wallet: Option<Result<rocksdb::DB, rocksdb::Error>>,
+    /// Holds the wallets ids in plain text, and the wallets information encrypted with a password.
+    db: Arc<rocksdb::DB>,
 }
 
 impl Storage {
-    /// Create a new instance of the Storage actor.
-    ///
-    /// It receives a path where to store:
-    ///
-    /// 1. The wallets database, which contains metadata about all created wallets.
-    /// 2. Each individual wallet database that is created.
-    pub fn new(db_path: PathBuf) -> Self {
-        let mut options = rocksdb::Options::default();
-
-        options.create_if_missing(true);
-
-        Self::with_options(db_path, options)
+    pub fn build<'a>() -> StorageBuilder<'a> {
+        StorageBuilder::default()
     }
 
-    /// Create a new instance of the Storage actor using the given database options.
-    pub fn with_options(db_path: PathBuf, options: rocksdb::Options) -> Self {
-        let wallets = rocksdb::DB::open(&options, db_path.join("wallets"));
+    pub fn new(db: Arc<rocksdb::DB>) -> Self {
+        Self { db }
+    }
 
-        Self {
-            db_path,
-            wallets,
-            wallet: None,
+    pub fn get_wallet_infos(&self) -> Result<Vec<wallet::WalletInfo>, error::Error> {
```

Comment:
What will happen when we have more than one key in storage? What about a generic private getter and public getters:

```rust
    pub fn get_wallet_infos(&self) -> Result<Vec<wallet::WalletInfo>, error::Error> { 
        self.get("wallet-infos")
    }
```

```rust
    fn get<V>(&self, key: &str) -> Result<V, error::Error> {
        let result = self
            .db
            .get(key)
            .map_err(error::Error::DbGetFailed)?;

        match result {
            Some(db_vec) => {
                let value =
                    deserialize(db_vec.as_ref()).map_err(error::Error::DeserializeFailed)?;
                Ok(value)
            }
            None => Ok(Vec::new()),
        }
    }
```

Reply:
I had it generic in the beginning but since depending on the key we will use encryption or not, I decided to strive for a simpler to understand solution like this. Also the internal type `DB` is already generic, so I think we would benefit more of a [Repository Pattern](https://martinfowler.com/eaaCatalog/repository.html) here

Reply:
I now understand what you meant, I added a `wallet/src/storage` module with all the generic operations.